### PR TITLE
Write Hash Chiplet constraints in AirScript

### PR DESCRIPTION
**Pull request by Overcastan**
Addressing: #202 
This PR adds Hash Chiplet constraints.

### CODE REVIEW

**Comment by Overcastan** [comment]
This PR has the same problem as #191, that is described in [this comment](https://github.com/0xPolygonMiden/air-script/pull/191#issuecomment-1466039654)

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:22]: Should we refactor this as:
`let f_rpr = !k0`

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:23]: 1. Should we rewrite this as:
`let f_bp = k2 & s[0] & !s[1] & !s[2]`

2. Should we declare an intermediate variable for `k2 & s[0]`?

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:31]: Should we declare an intermediate variable for `k0 & s[0]`?

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:30]: Should we rewrite this as:
`let f_out_next = k1 & !s[0]' & !s[1]'`

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:69]: This doesn't seem correct as there is no rhs?

**Review by tohrnii — CHANGES_REQUESTED** [review]
Thank you @Overcastan. I left a few comments.

I am wondering if we should allow logical operators for in intermediate variables and for intermediate variables? For e.g., should the following be valid?:

```
let a = s0 & !s1
let b = !a
```

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air:69]: Good catch! It should be equal 0. Thank you!

**Comment by Overcastan** [comment]
Allowing logical operators in intermediate variables looks really useful, but won't this complicate the process of parsing and processing expressions?

**Comment by grjte** [comment]
> Allowing logical operators in intermediate variables looks really useful, but won't this complicate the process of parsing and processing expressions?

I think this is a good idea and would be useful, but let's not do it now. Can we create an issue and put it in the "Nice to have" section of [this tracking issue](https://github.com/0xPolygonMiden/air-script/issues/162)?

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air]: Since these are constraints rather than variables, can we make use of selectors to do the following?

```
enf s[1]' = s[1] when !f_out & !f_out_next
enf s[2]' = s[2] when !f_out & !f_out_next
```

In fact, it might be best to make an `is_copied` function, because we check for that a lot in this module and you could use it elsewhere as well


```
enf is_copied([s[1]]) when !f_out & !f_out_next
enf is_copied([s[2]]) when !f_out & !f_out_next
```

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air]: We can use `is_binary` here, right?

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air]: We could use `is_copied` and a selector here, which I think improves readability:

```
enf is_copied([i]) when 1 - f_an - f_out
```

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air]: Note for the future - it might be nice to be able to combine constraint comprehension with selectors, e.g.

```
enf is_copied([h[j]]) for j in 0..4 when f_abp
```

What do you both think? (Or will this already be possible without extra work?)

**Comment by tohrnii** [comment]
> I think this is a good idea and would be useful, but let's not do it now. Can we create an issue and put it in the "Nice to have" section of this tracking issue?

Added #217

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air]: A few points here:
- I think using a selector for the absorption flag would improve readability (`when f_mp + f_mv + f_mu`)
- I think using `is_copied` would improve readability (e.g. `is_copied([h[j+4]])`)
- I think using selectors for `b` and `!b` would improve readability. This would mean splitting it into 2 constraints using a `match enf`, one of which is enforced `when !b & (f_mp + f_mv + f_mu)` and the other of which is enforced `when b & (f_mp + f_mv + f_mu)`.
- It needs to be applied to the entire digest, but it looks like it's only being applied to the first element of the digest right now

File: constraints/miden-vm/hash.air
Comment (tohrnii) [reply constraints/miden-vm/hash.air]: Currently we only allow identifiers as selectors and not expressions like `1 - f_an - f_out`. Do you think we should allow expressions as selectors? (To me it doesn't look very clean but not a strong opinion). Maybe we could define another intermediate variable for `1 - f_an - f_out` and use that.

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air:1]: My thought was that this file contains constraints for general hash operations, that's why I thought that it's possible to name it just `HashAir`, but `HashChipletAir` better displays the essence of the entire script, so I agree with you, we should rename it to `HashChipletAir`.

File: constraints/miden-vm/hash.air
Comment (tohrnii) [reply constraints/miden-vm/hash.air]: @grjte I think Bobbin suggested to write this as

```
ev foo(main: h[4]):
    enf is_copied(h) for h in h
enf foo(h[4]) when f_abp
```
instead.


Also, should we write this as:
```
enf is_copied([h]) for h in h[0..4]
```

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air]: As I understood, unfortunately, no. We should pass only columns in `is_binary` evaluator, and b is a variable. But it would be great if I got it wrong.

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air]: I'm not sure I understand what do you mean by digest in this context. Can you explain in more detail?

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air:58]: I'm not sure why should we create intermediate variables for the `k0 * (1 - s[0])` and similar expressions. I think in this case readability will be decreased since every row is described with set of selector values (e.g. `# this flag is set to 1 when flags are (i, j, k) and ... `) and in this case it will be more difficult to keep track of the value of the first flag

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:30]: `f_hout` and `f_sout` are undefined.

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:32]: `f_hout` and `f_sout` are undefined.

nit: I'd add a blank line. I think by convention we should have a blank line between a constraint and if the next line is a comment. Wdyt?

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:71]: `f_hout` and `f_sout` are undefined.

**Review by tohrnii — CHANGES_REQUESTED** [review]
Thank you @Overcastan, I just left a few comments. There is a recent change to logical operators and now they can be used in intermediate variables. Can you please refactor the code to incorporate that?

So, for example, something like:
```
let f_mpa = k0 * s[0] * binary_not(s[1]) * s[2]
```
can be re-written as
```
let f_mpa = k0 & s[0] & !s[1] & s[2]
```
Can you please also check in the memory chiplets PR if these changes need to be made?

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air]: > enf is_copied([h]) for h in h[0..4]
Yes, thank you.

> ev foo(main: h[4]):
>    enf is_copied(h) for h in h
> enf foo(h[4]) when f_abp

Yep, this is fine

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air]: The digest is the 4 field elements that make up the hash

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air:32]: Yes, I think we should always do this. I left a similar comment previously

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air]: > Maybe we could define another intermediate variable for 1 - f_an - f_out and use that.

yep, I think this would be better

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air:89]: Hmm you're right. I was thinking we'd use an is_binary function (rather than an evaluator), but then we still have to check something like `is_binary(b) = 0` which is probably less clear.

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:131]: I don't think we need to wait, since this isn't going into the tests. I would just change this now:

 ```
 match enf:
     is_copied(h[j + 4]) for j in 0..4 when !b & f_abs_node
     h[j + 8]' = h[j + 4] for j in 0..4 when b & f_abs_node
 ```

File: constraints/miden-vm/hash.air
Comment (grjte) [reply constraints/miden-vm/hash.air]: I think you're imagining this in a different way from what we intended. If you were to follow this suggestion, it would be given a meaningful name, so we would have something like the following:

```
let last_row_and_s0 = k0 * s[0]
let first_row_and_s0 = k2 * s[0]
```

But looking at this, I wonder if the most readable thing to do would actually be to rename the periodic columns like:

```
periodic_columns:
    cycle_row_7: [0, 0, 0, 0, 0, 0, 0, 1]
    cycle_row_6: [0, 0, 0, 0, 0, 0, 1, 0]
    cycle_row_0: [1, 0, 0, 0, 0, 0, 0, 0]
```

**Review by grjte — CHANGES_REQUESTED** [review]
Looks good, thanks @Overcastan. In addition to the things @tohrnii mentioned, I left a few things inline.

Also more generally - the comments on the flags tell when they're set, but it would be really helpful to include the semantic meaning as well.

For example:

```
    # Set to 1 when selector flags are (1,0,0) on rows which are 1 less than a multiple of 8.
    # Indicates the hash operation is absorption
    let f_abp = k0 * s[0] * binary_not(s[1]) * binary_not(s[2])
```

If you're not sure of the semantic meaning of the flags, feel free to message me

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air:20]: I agree, I put function on the top in the Memory Chiplet for the same reason. Not sure about additional headers though. On the one hand functions and evaluators are very similar, so this separation is not needed. But on the other hand if we have a lot of functions and evaluators, it will be very convenient to have this separation. I'll make this separation for now, let's see how it will look like.

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:72]: nit: There are typos here. It should be "Flag f_out_next is set to 1..."

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:70]: Should we make a helper function for `f_out` as well? Then this comment can go above the function definition and this evaluator and the ones below can be much smaller

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:160]: Similar to my previous comment - let's put all of these comments above the function definitions and remove them from here. That way this can be just 6 lines and this evaluator will be much easier to read. Let's also do that anywhere else any of these `get_f_*` functions are called

**Review by grjte — CHANGES_REQUESTED** [review]
Looks good, thank you! Just a couple of final comments about comments for the functions, then it will be ready to merge

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:177]: nit: I would prefer to move this up a few rows so it's declared with the other flags at the top of the evaluator

File: constraints/miden-vm/hash.air
Comment (grjte) [inline constraints/miden-vm/hash.air:90]: Since we already have this above the `get_f_out` function, I think we can remove it and just keep the line below

**Review by grjte — APPROVED** [review]
Looks great, thank you! I've left 2 final nits inline that I think it would be great to fix before merging, but they're not blocking. I'll leave it up to you to merge and you can make those updates first if you agree with them

File: constraints/miden-vm/hash.air
Comment (tohrnii) [inline constraints/miden-vm/hash.air:21]: nit: should we use logical operators here and other flag calculations as well?

**Review by tohrnii — APPROVED** [review]
Thank you @Overcastan, looks good to me. I just left one nit to write flag computation in terms of logical operators if you agree with the change.

File: constraints/miden-vm/hash.air
Comment (Overcastan) [reply constraints/miden-vm/hash.air:21]: Yes, I think we should do that. Thank you!

**Pull request merged**
